### PR TITLE
[FEATURE] Add HtmlPruner::removeRedundantClasses

### PR DESCRIPTION
**Pull request by JakeQZ**
Part of #380.

### CODE REVIEW

File: src/Emogrifier/HtmlProcessor/HtmlPruner.php [hunk 1]
```diff
@@ -41,4 +41,47 @@ public function removeInvisibleNodes()
 
         return $this;
     }
+
+    /**
```
Comment (oliverklee) [inline src/Emogrifier/HtmlProcessor/HtmlPruner.php]: By the way, would you like to add yourself as an `@author` to the class and the test case?

File: src/Emogrifier/HtmlProcessor/HtmlPruner.php [hunk 2]
```diff
@@ -41,4 +41,47 @@ public function removeInvisibleNodes()
...
+     * Removes classes that are no longer required (e.g. because there are no longer any CSS rules that reference them)
+     * from `class` attributes.
+     *
+     * Note that this does not inspect the CSS, but expects to be provided with a list of classes that are still in use.
+     *
+     * This method also has the (presumably beneficial) side-effect of minifying (removing superfluous whitespace from)
+     * `class` attributes.
+     *
+     * @param string[] $classesToKeep list of class names that should not be removed
```
Comment (oliverklee) [inline src/Emogrifier/HtmlProcessor/HtmlPruner.php]: The "list of" is redundant here.

File: src/Emogrifier/HtmlProcessor/HtmlPruner.php [hunk 3]
```diff
@@ -41,4 +41,47 @@ public function removeInvisibleNodes()
...
+                if ($nodeClassesToKeep !== []) {
+                    $node->setAttribute('class', \implode(' ', $nodeClassesToKeep));
+                } else {
+                    $node->removeAttribute('class');
+                }
+            }
+        } else {
+            foreach ($nodesWithClassAttribute as $node) {
+                $node->removeAttribute('class');
+            }
```
Comment (oliverklee) [inline src/Emogrifier/HtmlProcessor/HtmlPruner.php]: I propose we move both branches into to well-named methods.

File: src/Emogrifier/HtmlProcessor/HtmlPruner.php [hunk 4]
```diff
@@ -41,4 +41,47 @@ public function removeInvisibleNodes()
...
+     * Note that this does not inspect the CSS, but expects to be provided with a list of classes that are still in use.
+     *
+     * This method also has the (presumably beneficial) side-effect of minifying (removing superfluous whitespace from)
+     * `class` attributes.
+     *
+     * @param string[] $classesToKeep list of class names that should not be removed
+     *
+     * @return self fluent interface
+     */
+    public function removeRedundantClasses(array $classesToKeep)
```
Comment (oliverklee) [inline src/Emogrifier/HtmlProcessor/HtmlPruner.php]: Would it make sense to have give `$classesToKepp` a default value of `[]`?

File: src/Emogrifier/HtmlProcessor/HtmlPruner.php [hunk 5]
```diff
@@ -41,4 +41,47 @@ public function removeInvisibleNodes()
...
+    {
+        $nodesWithClassAttribute = $this->xPath->query('//*[@class]');
+
+        if ($classesToKeep !== []) {
```
Comment (oliverklee) [inline src/Emogrifier/HtmlProcessor/HtmlPruner.php]: Do we need this condition? Or is the second branch just a special case of the first one?

File: src/Emogrifier/HtmlProcessor/HtmlPruner.php [hunk 6]
```diff
@@ -41,4 +41,47 @@ public function removeInvisibleNodes()
...
+            // https://stackoverflow.com/questions/6329211/php-array-intersect-efficiency
+            // It's more efficient to invert the array and use `array_intersect_key()` when doing many intersections.
+            $classesToKeepAsKeys = \array_flip($classesToKeep);
+
+            foreach ($nodesWithClassAttribute as $node) {
+                $nodeClasses = \preg_split('/\\s++/', \trim($node->getAttribute('class')));
```
Comment (oliverklee) [inline src/Emogrifier/HtmlProcessor/HtmlPruner.php]: What does the second plus sign mean?

File: src/Emogrifier/HtmlProcessor/HtmlPruner.php [hunk 7]
```diff
@@ -41,4 +41,47 @@ public function removeInvisibleNodes()
...
+                $nodeClassesToKeep = \array_flip(\array_intersect_key(
+                    \array_flip($nodeClasses),
+                    $classesToKeepAsKeys
+                ));
```
Comment (oliverklee) [inline src/Emogrifier/HtmlProcessor/HtmlPruner.php]: I find this hard to grok with the multiple `array_flip`s. Maybe we can wrap this in a well-named method to reduce the level of detail required to understand this?

File: tests/Unit/Emogrifier/HtmlProcessor/HtmlPrunerTest.php [hunk 8]
```diff
@@ -85,4 +85,255 @@ public function removeInvisibleNodesRemovesNodesWithDisplayNone($displayNone)
...
+    }
+
+    /**
+     * @return string[][][]
+     */
+    public function classesToKeepDataProvider()
+    {
+        return [
+            'no classes to keep' => [[]],
+            '1 class to keep' => [['foo']],
```
Comment (oliverklee) [inline tests/Unit/Emogrifier/HtmlProcessor/HtmlPrunerTest.php]: To cover all equivalence classes, we'll also need a test case for 2 classes to keep (for the minimum version of "multiple classes").

File: tests/Unit/Emogrifier/HtmlProcessor/HtmlPrunerTest.php [hunk 9]
```diff
@@ -85,4 +85,255 @@ public function removeInvisibleNodesRemovesNodesWithDisplayNone($displayNone)
...
+        $html = '<p style="color: green;">hello</p>';
+        $subject = HtmlPruner::fromHtml('<html>' . $html . '</html>');
+
+        $subject->removeRedundantClasses($classesToKeep);
+
+        self::assertContains($html, $subject->render());
+    }
+
+    /**
+     * @return (string|string[])[][]
```
Comment (oliverklee) [inline tests/Unit/Emogrifier/HtmlProcessor/HtmlPrunerTest.php]: I propose making this more redundant, but easier to read:
`string[][]|string[][][]`

File: tests/Unit/Emogrifier/HtmlProcessor/HtmlPrunerTest.php [hunk 10]
```diff
@@ -85,4 +85,255 @@ public function removeInvisibleNodesRemovesNodesWithDisplayNone($displayNone)
...
+    {
+        $subject = HtmlPruner::fromHtml('<html>' . $html . '</html>');
+
+        $subject->removeRedundantClasses($classesToKeep);
+
+        self::assertNotContains('class', $subject->render());
+    }
+
+    /**
+     * @return (string|string[])[][]
```
Comment (oliverklee) [inline tests/Unit/Emogrifier/HtmlProcessor/HtmlPrunerTest.php]: Ditto.

File: tests/Unit/Emogrifier/HtmlProcessor/HtmlPrunerTest.php [hunk 11]
```diff
@@ -85,4 +85,255 @@ public function removeInvisibleNodesRemovesNodesWithDisplayNone($displayNone)
...
+     */
+    public function nonMatchedClassesDataProvider()
+    {
+        return [
+            '1 attribute, 1 class, no classes to keep' => [
+                'html' => '<p class="foo">hello</p>',
+                'classes to keep' => [],
+                'classes expected to be removed' => ['foo'],
```
Comment (oliverklee) [inline tests/Unit/Emogrifier/HtmlProcessor/HtmlPrunerTest.php]: I like having these descriptive sub-array keys.

File: tests/Unit/Emogrifier/HtmlProcessor/HtmlPrunerTest.php [hunk 12]
```diff
@@ -85,4 +85,255 @@ public function removeInvisibleNodesRemovesNodesWithDisplayNone($displayNone)
...
+
+    /**
+     * @test
+     *
+     * @param string $html
+     * @param string[] $classesToKeep
+     *
+     * @dataProvider nonMatchedClassesDataProvider
+     */
+    public function removeRedundantClassesRemovesClassAttributesContainingNoClassesToKeep($html, array $classesToKeep)
```
Comment (oliverklee) [inline tests/Unit/Emogrifier/HtmlProcessor/HtmlPrunerTest.php]: I find it confusing that the test has only two parameters, but the data provider provides three. I know that the DP gets recycling in another place, and don't have a good idea to solve this at the moment, though.

File: tests/Unit/Emogrifier/HtmlProcessor/HtmlPrunerTest.php [hunk 13]
```diff
@@ -85,4 +85,255 @@ public function removeInvisibleNodesRemovesNodesWithDisplayNone($displayNone)
...
+            ],
+            'last class in attribute is to be removed' => [
+                'html' => '<p class="foo bar baz">hello</p>',
+                'classes to keep' => ['foo', 'bar'],
+                'classes expected to be removed' => ['baz'],
+            ],
+        ];
+        $matchedClassesWithExtraWhitespaceData = $this->matchedClassesWithExtraWhitespaceDataProvider();
+
+        return $nonMatchedClassesData + $matchedClassesData + $matchedClassesWithExtraWhitespaceData;
```
Comment (oliverklee) [inline tests/Unit/Emogrifier/HtmlProcessor/HtmlPrunerTest.php]: Instead of having a data provider with a different kinds of data sets, I propose we have three separate tests that use each data subset and focus on one aspect of the behavior.

File: tests/Unit/Emogrifier/HtmlProcessor/HtmlPrunerTest.php [hunk 14]
```diff
@@ -85,4 +85,255 @@ public function removeInvisibleNodesRemovesNodesWithDisplayNone($displayNone)
...
+                'classes expected to be removed' => ['bar'],
+            ],
+            '1 attribute, 2 classes separated by newline, both to be kept' => [
+                'html' => "<p class=\"foo\nbar\">hello</p>",
+                'classes to keep' => ['foo', 'bar'],
+            ],
+            '1 attribute, 2 classes separated by newline, 1 to be kept' => [
+                'html' => "<p class=\"foo\nbar\">hello</p>",
+                'classes to keep' => ['foo'],
+                'classes expected to be removed' => ['bar'],
```
Comment (oliverklee) [inline tests/Unit/Emogrifier/HtmlProcessor/HtmlPrunerTest.php]: Instead of relying on the default values for the parameters, please let's make the third value explicit. Then the reader does not need to know the test method interface to understand the data from the data provider.

File: tests/Unit/Emogrifier/HtmlProcessor/HtmlPrunerTest.php [hunk 15]
```diff
@@ -85,4 +85,255 @@ public function removeInvisibleNodesRemovesNodesWithDisplayNone($displayNone)
...
+                    $expectedInstanceCount,
+                    \substr_count($result, $class),
+                    'asserting \'' . $result . '\' contains ' . $expectedInstanceCount . ' instance(s) of "' . $class
+                    . '"'
+                );
+            }
+        }
+        foreach ($classesExpectedToBeRemoved as $class) {
+            self::assertNotContains($class, $result);
+        }
```
Comment (oliverklee) [inline tests/Unit/Emogrifier/HtmlProcessor/HtmlPrunerTest.php]: Please let's split this up into separate tests.

File: src/Emogrifier/HtmlProcessor/HtmlPruner.php (hunk 1 above)
Comment (JakeQZ) [reply src/Emogrifier/HtmlProcessor/HtmlPruner.php]: Sure, will do 😀

File: src/Emogrifier/HtmlProcessor/HtmlPruner.php (hunk 2 above)
Comment (JakeQZ) [reply src/Emogrifier/HtmlProcessor/HtmlPruner.php]: OK, will remove it.

File: src/Emogrifier/HtmlProcessor/HtmlPruner.php (hunk 3 above)
Comment (JakeQZ) [reply src/Emogrifier/HtmlProcessor/HtmlPruner.php]: OK, will refactor...

File: src/Emogrifier/HtmlProcessor/HtmlPruner.php (hunk 4 above)
Comment (JakeQZ) [reply src/Emogrifier/HtmlProcessor/HtmlPruner.php]: Yes, I think so...

File: src/Emogrifier/HtmlProcessor/HtmlPruner.php (hunk 5 above)
Comment (JakeQZ) [reply src/Emogrifier/HtmlProcessor/HtmlPruner.php]: > Or is the second branch just a special case of the first one?

It is indeed.  The first branch would achieve the same result, but do a whole lot of work for nothing - i.e. for every class attribute, doing a `preg_split`, computing the intersection with an empty array, testing that the result is empty (it of course always will be), then finally deducing that the attribute should be removed.

File: src/Emogrifier/HtmlProcessor/HtmlPruner.php (hunk 6 above)
Comment (JakeQZ) [reply src/Emogrifier/HtmlProcessor/HtmlPruner.php]: It's the [possessive quantifier](https://www.regular-expressions.info/possessive.html) - it tells the regex engine not to worry about backtracking, so makes it more optimal and helps mitigate against [catastrophic failure](https://www.regular-expressions.info/catastrophic.html).

If the regex was instead `/\s+\n/`, the possessive quantifier could not be used, as the regex engine may need to backtrack to see if one of the whitespace characters matched was in fact a newline - however that regex could be rewritten as `/[^\S\n]++\n/` to ensure the first part won't match a newline, then the possessive quantifier can be used.  Other examples include rewriting `/\(.*\)/` as `/\([^\)]*+\)` (matching something in brackets) - with `/\(.*\)/` the regex engine could have to backtrack a long way to find the closing bracket (or lack of).

File: src/Emogrifier/HtmlProcessor/HtmlPruner.php (hunk 7 above)
Comment (JakeQZ) [reply src/Emogrifier/HtmlProcessor/HtmlPruner.php]: Or perhaps a simple class could be provided to abstract the detail, so we could have

```php
            $classesToKeepIntersector = new BatchArrayIntersector($classesToKeep);
            ....
                $nodeClassesToKeep = $classesToKeepIntersector->intersectWith($nodeClasses);
```

I thought somebody might have already written one, but [could not find anything](https://www.google.com/search?q=github+simple+php+class+for+batch+array+intersection+using+flipped+arrays+and+array_intersect_key+instead+of+array_intersect).

File: tests/Unit/Emogrifier/HtmlProcessor/HtmlPrunerTest.php (hunk 8 above)
Comment (JakeQZ) [reply tests/Unit/Emogrifier/HtmlProcessor/HtmlPrunerTest.php]: OK, will add it.

File: tests/Unit/Emogrifier/HtmlProcessor/HtmlPrunerTest.php (hunk 9 above)
Comment (JakeQZ) [reply tests/Unit/Emogrifier/HtmlProcessor/HtmlPrunerTest.php]: That wouldn't be correct - it would mean that either all elements of the 2nd level array are strings, or they're all arrays of strings.  Whereas in fact, some of them are strings and some of them are arrays of strings.  The only viable alternative I can think of would be `mixed[][]`...

File: tests/Unit/Emogrifier/HtmlProcessor/HtmlPrunerTest.php (hunk 11 above)
Comment (JakeQZ) [reply tests/Unit/Emogrifier/HtmlProcessor/HtmlPrunerTest.php]: I thought it might be possible, and would help readability, particularly with two arrays of classes.  So I tried it and found it was possible 😎

Though as they are for human consumption only, `'html'` should really be `'HTML'`.  I'll change that...

File: tests/Unit/Emogrifier/HtmlProcessor/HtmlPrunerTest.php (hunk 12 above)
Comment (JakeQZ) [reply tests/Unit/Emogrifier/HtmlProcessor/HtmlPrunerTest.php]: From memory, if we add an unused parameter, PHPCS (or PHPMD) will complain.

File: tests/Unit/Emogrifier/HtmlProcessor/HtmlPrunerTest.php (hunk 13 above)
Comment (JakeQZ) [reply tests/Unit/Emogrifier/HtmlProcessor/HtmlPrunerTest.php]: OK, I'll look into reworking it...

File: tests/Unit/Emogrifier/HtmlProcessor/HtmlPrunerTest.php (hunk 14 above)
Comment (JakeQZ) [reply tests/Unit/Emogrifier/HtmlProcessor/HtmlPrunerTest.php]: OK

File: tests/Unit/Emogrifier/HtmlProcessor/HtmlPrunerTest.php (hunk 15 above)
Comment (JakeQZ) [reply tests/Unit/Emogrifier/HtmlProcessor/HtmlPrunerTest.php]: OK.  We can have `removeRedundantClassesRemovesClassesNotToKeep` and `removeRedundantClassesNotRemovesClassesNotToKeep`.  It may then be realized that not all the datasets are applicable to one or other, which might help with

> Instead of having a data provider with a different kinds of data sets, I propose we have three separate tests that use each data subset and focus on one aspect of the behavior.

**Comment by JakeQZ** [comment]
I've just replied to the comments for now.  I'll look at committing some changes tomorrow... 🌙

**Comment by JakeQZ** [comment]
> I find this hard to grok with the multiple `array_flip`s. Maybe we can wrap this in a well-named method to reduce the level of detail required to understand this?

> Or perhaps a simple class could be provided to abstract the detail

On the basis of the adage “if a job’s worth doing, it’s worth doing well”, I’ve added such a class.  I dropped the prefix `Batch…` as that may imply a method that does the batching, whereas in fact it has a method intended to be called repeatedly as part of some batch processing.  See #710.

File: src/Emogrifier/HtmlProcessor/HtmlPruner.php (hunk 5 above)
Comment (JakeQZ) [reply src/Emogrifier/HtmlProcessor/HtmlPruner.php]: I'm adding a comment to the code clarifying this.

**Comment by JakeQZ** [comment]
I've made changes based on the review suggestions and repushed.

**Comment by JakeQZ** [comment]
> I've made changes based on the review suggestions and repushed.

In this change I introduced these private methods:

```php
private function removeClassesFromNodes(\DOMNodeList $nodeList, array $classesToKeep);
private function removeClassAttributeFromNodes(\DOMNodeList $nodeList);
```

I chose `…Nodes` in the method name because the parameter is a `DOMNodeList`, but now I think about it, the methods should probably be named `…Elements` because they require that the nodes are elements (only elements can have attributes).  And perhaps then the parameter should be `$elements` rather than `$nodeList`, and the local variables renamed `…node…` to `…element…`...?

(We also have various instances in the code base of private methods and local variables using ‘node’ in the name, where they are necessarily elements, and perhaps should consider changing the names for clarity, but I’d say not for 3.0.0 – except #717.)

**Comment by oliverklee** [comment]
> but now I think about it, the methods should probably be named `…Elements` because they require that the nodes are elements (only elements can have attributes). And perhaps then the parameter should be `$elements` rather than `$nodeList`, and the local variables renamed `…node…` to `…element…...`?

Yes, this makes sense and sounds good.

**Comment by JakeQZ** [comment]
> > the methods should probably be named `…Elements` … the parameter should be `$elements` rather than `$nodeList`, and the local variables renamed `…node…` to `…element…...`?
> 
> Yes, this makes sense and sounds good.

I've renamed accordingly and repushed.

File: src/Emogrifier/HtmlProcessor/HtmlPruner.php [hunk 16]
```diff
@@ -41,4 +44,69 @@ public function removeInvisibleNodes()
 
         return $this;
     }
+
+    /**
+     * Removes classes that are no longer required (e.g. because there are no longer any CSS rules that reference them)
```
Comment (oliverklee) [inline src/Emogrifier/HtmlProcessor/HtmlPruner.php]: Should there be a comma after "e.g."?

File: src/Emogrifier/HtmlProcessor/HtmlPruner.php [hunk 17]
```diff
@@ -41,4 +44,69 @@ public function removeInvisibleNodes()
...
+     * from `class` attributes.
+     *
+     * Note that this does not inspect the CSS, but expects to be provided with a list of classes that are still in use.
+     *
+     * This method also has the (presumably beneficial) side-effect of minifying (removing superfluous whitespace from)
+     * `class` attributes.
+     *
+     * @param string[] $classesToKeep class names that should not be removed
```
Comment (oliverklee) [inline src/Emogrifier/HtmlProcessor/HtmlPruner.php]: Nitpick: class names -> names of classes

File: tests/Unit/Emogrifier/HtmlProcessor/HtmlPrunerTest.php [hunk 18]
```diff
@@ -85,4 +86,281 @@ public function removeInvisibleNodesRemovesNodesWithDisplayNone($displayNone)
...
+
+        $result = $subject->render();
+        foreach ($classesToKeep as $class) {
+            $expectedInstanceCount = \substr_count($html, $class);
+            self::assertSame(
+                $expectedInstanceCount,
+                \substr_count($result, $class),
+                'asserting \'' . $result . '\' contains ' . $expectedInstanceCount . ' instance(s) of "' . $class
+                . '"'
+            );
```
Comment (oliverklee) [inline tests/Unit/Emogrifier/HtmlProcessor/HtmlPrunerTest.php]: I propose we move this into a well-name static `assert*` method.

File: tests/Unit/Emogrifier/HtmlProcessor/HtmlPrunerTest.php [hunk 19]
```diff
@@ -85,4 +86,281 @@ public function removeInvisibleNodesRemovesNodesWithDisplayNone($displayNone)
...
+     */
+    public function removeRedundantClassesMinifiesClassAttributes($html, array $classesToKeep)
+    {
+        $subject = HtmlPruner::fromHtml('<html>' . $html . '</html>');
+
+        $subject->removeRedundantClasses($classesToKeep);
+
+        \preg_match_all('/class="([^"]*+)"/', $subject->render(), $classAttributeMatches);
+        foreach ($classAttributeMatches[1] as $classAttributeValue) {
+            self::assertNotRegExp('/^\\s|\\s{2}|\\s$/', $classAttributeValue);
```
Comment (oliverklee) [inline tests/Unit/Emogrifier/HtmlProcessor/HtmlPrunerTest.php]: Can we move this into a static `assert*` method with a descriptive name?

File: src/Emogrifier/HtmlProcessor/HtmlPruner.php (hunk 16 above)
Comment (JakeQZ) [reply src/Emogrifier/HtmlProcessor/HtmlPruner.php]: There's some discussion on [StackExchange](https://english.stackexchange.com/questions/16172/should-i-always-use-a-comma-after-e-g-or-i-e) that suggests it may be a matter of style, but also an [answer with an example in brackets](https://english.stackexchange.com/questions/16172/should-i-always-use-a-comma-after-e-g-or-i-e#answer-16188) (as here) where there is no comma, and also a [reference link showing a comma only when the example is not in brackets](https://theoatmeal.com/comics/ie).  So I'll leave it as it is.

File: src/Emogrifier/HtmlProcessor/HtmlPruner.php (hunk 17 above)
Comment (JakeQZ) [reply src/Emogrifier/HtmlProcessor/HtmlPruner.php]: Done.

File: tests/Unit/Emogrifier/HtmlProcessor/HtmlPrunerTest.php (hunk 18 above)
Comment (JakeQZ) [reply tests/Unit/Emogrifier/HtmlProcessor/HtmlPrunerTest.php]: Done.

File: tests/Unit/Emogrifier/HtmlProcessor/HtmlPrunerTest.php (hunk 19 above)
Comment (JakeQZ) [reply tests/Unit/Emogrifier/HtmlProcessor/HtmlPrunerTest.php]: Done.

**Comment by JakeQZ** [comment]
> Just some minor remarks.

Have made suggested changes.

> And could you please rebase?

Well, I used `git rebase master` and it seems to have gone somewhat awry.  Initially I had not done a `git pull` on `master` so it was not up to date.  But after doing so, it kept trying to merge the initial changes from this PR on top of the later changes.  When finally all sorted, I see that the file diff now shows all of the other changes from `master` since this PR was started - exactly what we were trying to avoid!

I've saved the command prompt transcript which I can post to identify what I might have done wrong...

**Comment by oliverklee** [comment]
> Well, I used `git rebase master` and it seems to have gone somewhat awry

I'll do my best to help.

**Comment by oliverklee** [comment]
Some ideas to resolve the git problem:
1. change to master and pull (just to be on the safe side)
1. drop the merge commits from the branch (using `git rebase -i HEAD~25`
1. rebase from master again and resolve any rebase conflicts in the individual commits

**Comment by oliverklee** [comment]
@JakeQZ And if you like, I can also resolve the problem on my machine, force-push and then provide the steps I executed.

**Comment by oliverklee** [comment]
Should this close #380 when merged?

**Comment by JakeQZ** [comment]
>     2. drop the merge commits from the branch (using `git rebase -i HEAD~25` 
>     3. rebase from master again and resolve any rebase conflicts in the individual commits

I think I've sorted it.  It seems that after `git rebase master`, the local copy of the branch must be `--force push`ed.  If an ordinary `push` is attempted, you're told you need to do a `pull` first which then ends up with the entire set of commits on the branch being duplicated (or triplicated), and the 'patching base' not being moved at all.

**Comment by JakeQZ** [comment]
> Should this close #380 when merged?

Not yet - there's the final method which takes a `CssInliner` parameter still to do...

**Comment by oliverklee** [comment]
>  If an ordinary `push` is attempted, you're told you need to do a `pull` first which then ends up with the entire set of commits on the branch being duplicated (or triplicated), and the 'patching base' not being moved at all.

Oh, yes, that message. The action the message recommends is completely bogus. Just last week, I told exactly that to the attendees at a Git workshop. :-)

File: tests/Unit/Emogrifier/HtmlProcessor/HtmlPrunerTest.php [hunk 20]
```diff
@@ -85,4 +86,306 @@ public function removeElementsWithDisplayNoneRemovesElementsWithDisplayNone($dis
...
+    /**
+     * Asserts that the number of occurrences of `$needle` within the string `$haystack` is as expected.
+     *
+     * @param int $expectedCount
+     * @param string $haystack
+     * @param string $needle
+     *
+     * @throws \PHPUnit_Framework_AssertionFailedError
+     */
+    public static function assertSubstringCount($expectedCount, $haystack, $needle)
```
Comment (oliverklee) [inline tests/Unit/Emogrifier/HtmlProcessor/HtmlPrunerTest.php]: The two helper functions don't need to be public; private should suffice here.

Comment (JakeQZ) [reply tests/Unit/Emogrifier/HtmlProcessor/HtmlPrunerTest.php]: OK, I've made them private and repushed.
